### ISSUE TITLE
Questions about mel spectrogram feature extraction

### ISSUE THREAD

**Issue by Btwestyo** [MihawkHu/DCASE2020_task1#19]
On L36: https://github.com/MihawkHu/DCASE2020_task1/blob/master/task1a/10class/extr_feat_2020_nodelta_scaled.py#L36 why do you add 1.e-8 to the spectrogram features and then take the log of it?

Also from L39-L41 of that same file I see you also normalize the features between the min and max of the features. Is this necessary to do and what is the reason for doing this? 

best,
Brian

**Comment by MihawkHu** [comment]
Hi,

Thanks for your interest. We added 1.e-8 to the spectrogram features to avoid the possible clash when performing log operation over the zero value. We scaled and normalized each feature to between 0 and 1. You can refer to our report for the detail experimental results. https://arxiv.org/pdf/2007.08389.pdf. It's not necessary and it has better empirical results.

**Comment by Btwestyo** [comment]
Thanks. On L34: (https://github.com/MihawkHu/DCASE2020_task1/blob/master/task1a/10class/extr_feat_2020_nodelta_scaled.py#L34)

you essentially extract the spectrogram features (2D array of frequency vs time bins). For most of the audio files the dimensions are: 128 (frequency bins) x 431 (time bins). The problem for me is that I am trying to store all of the features in a numpy array and it turns out that not all of the audio files have the same 2D dimensions, some of the files lead to different dimensions that are not 128 x 431. 

Any suggestion on how to deal with that? My solution has been to take a mean and collapse the entire time dimension:
`np.mean(librosa.feature.melspectrogram(stereo[:], sr=sr, n_fft=num_fft, hop_length=hop_length, n_mels=num_freq_bin, fmin=0.0, fmax=sr/2, htk=True, norm=None).T,axis=0) `

But by doing that I don't think it is a spectrogram anymore. Those reduced features seem to be okay, but surely I must be losing something?

**Comment by MihawkHu** [comment]
Hi, in our experiments we did not meet the problem you mentioned. The frequency bin is assigned by you so it should be equal to each other as 128. The time bin is corresponded to the total length of each audio clip. Cuz each audio clip is fixed to 1s in this data set, so it should be 431 for all if you use the same window length and hop length as us. Could you let me know which one you found has the different feature size?

**Comment by Btwestyo** [comment]
Hi,

I resolved the different feature size issue, apparently one of my wav files did not download correctly. I do have several questions regarding training and feature extraction. 

1. I see that for your testing data, you also compute the delta and delta-delta of the mel specgrogram for testing/validation only: https://github.com/MihawkHu/DCASE2020_task1/blob/f662825bf924677b0e4e38bc0b61985d8412b936/task1a/10class/resnet/train_resnet.py#L59

Do you train on these features as well or is it just for evaluation? Does evaluating/training with these additional channels make performance better? On L81 of the same file, you also say:
"# Due to the memory limitation, in the training stage we split the training data". I'm assuming I don't have to do those splits if I don't have memory limitations? 

2. You say in your paper that the final input tensor size is 423 x 128 x 3. I'm assuming those 2 extra channels are the deltas and delta deltas in the previous question? 

Right now, I am starting off simple with a 2 layer CNN architecture and 2 fully-connected layers. I am training on mel spectrograms of dim: 128 x 431 (10 million trainable parameters) with scaling between [0,1]. After about 20 epochs my training/test accuracy is completed stalled at 10% and I'm trying to debug whether my feature input of 128 x 431 is the problem and needs deltas/delta deltas, if I need to use your LR warm restart routine, or something else.

**Comment by MihawkHu** [comment]
For your first questions, the evaluation data is only used for evaluation, never for training the model. 
For your second question, you dont need to split the training data if you dont have the memory limitations on your machine. 
For your third question, yes the two extra channels are the delta and delta-delta. 
For your forth question, I'm not sure what's the problem with your training. It looks like the model does not get converged. You can double check the training settings, like optimizer, learning rate and input/output.

**Comment by Btwestyo** [comment]
Thanks Hu for the reply. I was able to fix convergence, it looks OK now. From your experience does the delta and delta-delta features help with improved accuracy over jut 1 channel (128 x 431)?

**Comment by MihawkHu** [comment]
In our experiments delta and delta-delta did bring some gains to model performance. But you can still use only 1 channel if you want to have smaller model input.

**Comment by Btwestyo** [comment]
Hu,

I am currently using the features from only 1 channel: 128 x 431 spectrogram and my performance on the test set is only 45%. I see that your paper, however, has 70% accuracy for the baseline Resnet model. Were you using any data augmentation, such as mixup to achieve that? If not, then it appears that the 3 input channels 128 x 431 x 3 is the only thing I'm missing to go from 45% accuracy to 70% accuracy. 

I see that in your (https://github.com/MihawkHu/DCASE2020_task1/blob/f662825bf924677b0e4e38bc0b61985d8412b936/task1a/10class/extr_feat_2020_nodelta_scaled.py) file, it only extracts and output features for 1 channel: 128 x 431 spectrogram without the extra features, hence the name "no delta scaled". 

If I understand correctly, are you getting the 2 extra channels using this routine?
https://github.com/MihawkHu/DCASE2020_task1/blob/f662825bf924677b0e4e38bc0b61985d8412b936/task1a/10class/resnet/DCASE_training_functions.py#L64

I'm a bit confused by that routine, it looks like it uses mixup data augmentation?

**Comment by MihawkHu** [comment]
Hi, 

To save the disk space we only extrated and dumped the logmel features without delta and delta-delta. The delta and delta-delta are computed on the fly during the training stage. You can refer to this line here, https://github.com/MihawkHu/DCASE2020_task1/blob/master/task1a/10class/fcnn/DCASE_training_functions.py#L98, where delta is computed. Yes this Generator class also carries out some data augmentation techniques that can be performned on the fly, including 
mixup: https://github.com/MihawkHu/DCASE2020_task1/blob/master/task1a/10class/fcnn/DCASE_training_functions.py#L147
random cropping: https://github.com/MihawkHu/DCASE2020_task1/blob/master/task1a/10class/fcnn/DCASE_training_functions.py#L137
spec-augment: https://github.com/MihawkHu/DCASE2020_task1/blob/master/task1a/10class/fcnn/DCASE_training_functions.py#L130.

**Comment by Btwestyo** [comment]
Hi,

Do you use mixup for all of your results? You had a bunch of columns in your paper referring to whether the data augmentation was used or not, but I didn't see a column for mixup so I wasn't sure which results included mixup or not.

-Brian

**Comment by MihawkHu** [comment]
Hi,

In our conference paper https://arxiv.org/pdf/2011.01447.pdf and challenge technique report https://arxiv.org/pdf/2007.08389.pdf, Mixup is used for every experimental items in result tables. Particularly we set mixup as a default method, and according to our experimental experience, Mixup is very important for the model training.

**Comment by Btwestyo** [comment]
I was re-reading your paper and if I understand correctly you're using the two-stage classifier for your results. I mistakenly though your results were for the 10-class classifier. 

Can you explain how your 3-class classifier and 10-class classifier give you a final accuracy score? Do you train a 3-class (indoor, outdoor, transportation) model and also train a 10-class model independently and combine scores in some way? Or for each audio file you use both classifiers or something?

**Comment by MihawkHu** [comment]
Hi,

Thanks for your interests in our paper. Particularly, we trained 10-class classifiers and 3-class classifiers independently. The codes are presented in https://github.com/MihawkHu/DCASE2020_task1/tree/master/task1a/10class, and https://github.com/MihawkHu/DCASE2020_task1/tree/master/task1a/3class. The training stages are very similar to each other. You can refer to our paper for details of how we performed the two-stage procedure. In our paper we also listed the detail results of single system and two-stage system. 

In the given evalution scripts, https://github.com/MihawkHu/DCASE2020_task1/blob/master/task1a/eval_hybrid.py,  it's from our hybrid two-stage system. We have provieded each pretrained single model, you can refer to the tensorflow1 version models here, https://github.com/MihawkHu/DCASE2020_task1/tree/master/task1a/10class/pretrained_models for 10-class classifiers, and here https://github.com/MihawkHu/DCASE2020_task1/tree/master/task1a/3class/pretrained_models for 10-class classifiers.

**Comment by Btwestyo** [comment]
Okay so if I understand correctly, for each test sample, you output prediction probabilities for the 10-class classifier and the 3-class classifier and then you multiply them in a certain way and then take the max probability to get the final prediction? I.e. getting the classification performance requires a bit of post-processing? 

Lastly, your CNN networks are they pre-trained on systems or you train them from scratch?

**Comment by MihawkHu** [comment]
Hi, 

Your understandings are totally correct. I'm not sure I fully understand your second question. Our provieded pre-trained models are CNN models trained from the scratch, based on the methods we discussed in our paper.
